File: 0021-Merge-Two-Sorted-Lists/0021.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2) {

    if(list1 == NULL && list2 != NULL) return list2;
    if(list1 != NULL && list2 == NULL) return list1;
    if(list1 == NULL && list2 == NULL) return NULL;

    struct ListNode *temp, *insertTemp;
    struct ListNode *otherList;
    struct ListNode *newList;

    if(list1->val <= list2->val) {
        newList = list1;
        temp = list1;
        otherList = list2;
    } else {
        newList = list2;
        temp = list2;
        otherList = list1;
    }


    while(temp != NULL) {

        if( otherList != NULL && 
            temp->next != NULL &&
            temp->val <= otherList->val &&
            otherList->val <= temp->next->val) {

                insertTemp = otherList->next;
                otherList->next = temp->next;
                temp->next = otherList;
                otherList = insertTemp;

            } else if(otherList != NULL && 
                      temp->next == NULL) {

                temp->next = otherList;
                break;

            } else {
                temp = temp->next;
            }
        }
        
    return newList;
}

File: 0021-Merge-Two-Sorted-Lists/0021.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

impl Solution {
    pub fn merge_two_lists(
        mut list1: Option<Box<ListNode>>,
        mut list2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {

        let mut temp = Some(Box::new(ListNode::new(0)));
        let mut tail = &mut temp;

        while list1.is_some() && list2.is_some() {

            let l1_val = list1.as_ref().unwrap().val;
            let l2_val = list2.as_ref().unwrap().val;

            if l1_val <= l2_val {
                let mut node = list1.take().unwrap();
                list1 = node.next.take();
                tail.as_mut().unwrap().next = Some(node);
            } else {
                let mut node = list2.take().unwrap();
                list2 = node.next.take();
                tail.as_mut().unwrap().next = Some(node);
            }

            tail = &mut tail.as_mut().unwrap().next;
        }

        tail.as_mut().unwrap().next = if list1.is_some() { list1 } else { list2 };

        temp.unwrap().next
    }
}
